File: common-rule-engine/src/test/files/checks/go/SwitchWithoutDefaultCheck/SwitchWithoutDefaultCheck.go
package main

type f struct {
	a, b, c int
}

func foo(tag int, x f) {
	switch tag {
	default: // Compliant
		bar()
	case 0, 1, 2, 3:
		bar()
	case 4, 5, 6, 7:
		bar()
	}


	switch tag {
	case 0, 1, 2, 3:
		bar()
	default:  // Noncompliant {{Move this default to the start or end of the switch.}}
		bar()
	case 4, 5, 6, 7:
		bar()
	}

	switch tag { // Noncompliant {{Add a default case to this switch.}}
	case 0, 1, 2, 3:
		bar()
	case 4, 5, 6, 7:
		bar()
	}

	switch tag { // Compliant
	case 0, 1, 2, 3:
		bar()
	case 4, 5, 6, 7:
		bar()
	default:
		bar()
	}

	switch tag { // Compliant
	case 0, 1, 2, 3:
		bar()
	case 4, 5, 6, 7:
		switch tag {  // Noncompliant {{Add a default case to this switch.}}
		case 5:
			bar()
		case 6:
			bar()
		}
	default:
		bar()
	}

	switch x.(type) { // Noncompliant {{Add a default case to this switch.}}
	case nil:
		bar("x is nil")
	case bool, string:
		bar("type is bool or string")
	}

	switch x.(type) {
	case nil:
		bar("x is nil")
	default:  // Noncompliant {{Move this default to the start or end of the switch.}}
		bar("don't know the type")
	case bool, string:
		bar("type is bool or string")
	}

	switch x.(type) {
	default:  // Compliant
		bar("don't know the type")
	case nil:
		bar("x is nil")
	case bool, string:
		bar("type is bool or string")
	}

	switch x.(type) { // Compliant
	case nil:
		bar("x is nil")
	case bool, string:
		bar("type is bool or string")
	default:
		bar("don't know the type")
	}
}

func bar(s ... string) {
}

File: common-rule-engine/src/test/files/checks/SwitchWithoutDefaultCheck/SwitchWithoutDefaultCheck.java
import java.time.DayOfWeek;

class Foo {
  void foo(MyEnum myEnum, DayOfWeek dow) {
    switch (0) { // Noncompliant {{Add a default case to this switch.}}
    }

    switch (0) { // Noncompliant
      case 0:
    }

    switch (0) { // Compliant
      default:
    }

    switch (0) {
      default:   // Compliant -->  order is handled by S4524
      case 0:
    }

    switch (myEnum) { // Noncompliant
      case A:
        break;
      case B:
        break;
    }

    // java:Compliant: Common rule engine not able to say that all cases of the enum are covered
    switch (myEnum) { // Noncompliant
      case A:
      case B:
      case C:
        break;
    }

    switch (myEnum) { // Compliant
      case A:
      case B:
        break;
      default:
        break;
    }

    switch (dow) { // Noncompliant
      case FRIDAY:
        break;
      case MONDAY:
        break;
    }

    switch (myEnum) { // Noncompliant
      case A:
      case B:
        switch (dow) { // Compliant
          case FRIDAY:
            break;
          default:
            break;
        }
        break;
    }

    switch (myEnum) {
      case A:
        break;
      default:  // Compliant - order is handled by S4524
        break;
      case B:
        break;
    }

    switch (myEnum) { // Compliant
      case A:
        switch (dow) { // Compliant
          case FRIDAY:
            break;
          default:
            break;
        }
        break;
      case B:
        break;
      default:
    }
  }
}

enum MyEnum {
  A, B, C;

  MyEnum field;
}
